### PR TITLE
Add _MatchingEngine and _StringProcessing modules.

### PR DESCRIPTION
**Pull request by rxwei**
These modules are part of the experimental declarative string processing feature. If accepted to the Standard Library, _StringProcessing will be available via implicit import just like _Concurrency, though _MatchingEngine will still be hidden as an implementation detail.

`_MatchingEngine` will contain the general-purpose pattern matching engine ISA, bytecode, and executor. `_StringProcessing` will contain regular expression and pattern matching APIs whose implementation depends on the matching engine..

Also consolidates frontend flag `-enable-experimental-regex` as `-enable-experimental-string-processing`.

Resolves rdar://85478647.

### CODE REVIEW

**Comment by rxwei** [comment]
@swift-ci please test

**Review by milseman — APPROVED** [review]
LGTM, after @benrimmington's fixes.

You might want to get this reviewed by someone more familiar with the CMake side though.

File: utils/refactor-check-compiles.py [hunk 1]
```diff
@@ -103,6 +103,8 @@ def main():
     extra_both_args = []
     if args.enable_experimental_concurrency:
         extra_both_args.append('-enable-experimental-concurrency')
+    if args.enable_experimental_concurrency:
```
[the remark this answers is not in the document]
Comment (rxwei) [reply utils/refactor-check-compiles.py:106]: Good catch!

**Comment by rxwei** [comment]
@swift-ci please test

**Review by edymtt — APPROVED** [review]
Looks good to me with regard to CMake/build system changes

**Comment by rxwei** [comment]
@swift-ci please clean test

**Comment by rxwei** [comment]
@swift-ci please clean test macos platform

**Comment by rxwei** [comment]
@swift-ci please clean test

**Comment by rxwei** [comment]
@swift-ci please clean test macOS platform

**Comment by rxwei** [comment]
@swift-ci please test

**Comment by rxwei** [comment]
@swift-ci please test

**Comment by rxwei** [comment]
@swift-ci please test and merge

**Comment by milseman** [comment]
@swift-ci please test MacOS platform

**Pull request merged**